toc.md

```diff
@@ -428,77 +429,232 @@ An additional backlog of 936 hours of image processing jobs everyday, along with
+#### Using Custom Algorithms
 
-With Sidekiq's random polling algorithm, it took 242 seconds. Our turn. We benchmarked with our evenly balancing algorithm with an even number of workers for each queue regardless of job types or queue load, and we stand at 357s. Much worse...
+For this optimization we decided to benchmark Workerholic against Sidekiq by using 10,000 non-blocking, cpu-blocking, and IO-blocking jobs.
 
-Next, we tried auto-balancing workers based on queue load, but still no assumption on jobs. Slightly better, but still not good enough. We went through a second iteration of auto-balancing where we identified IO-bound queues and CPU-bound queues, where we assign only one worker per CPU-bound queue and auto-balance the rest. As you may recall, having more workers on CPU-blocking jobs makes no difference, which is a waste of Workerholic's resources. With that we rang in at 223s, which is great! We call this algorithm the Adaptive and Successive Provisioning (ASP).
+![efficiency_algorithms_benchmark](/images/efficiency_algorithms_benchmark_1.png){:width="400"}
 
-*Note: we want to quickly mention that we did not build Workerholic to compete with Sidekiq, and that you should not prefer our library to theirs' or vice-versa. We just chose Sidekiq to benchmark against because it is the leader of background job processing in Ruby and we thought that is the bar we should aim for. Could we do better? Maybe, maybe not. We haven't tried to yet, because we were satisfied with the current results.*
+As shown in the above diagram, Sidekiq's random polling algorithm took 242 seconds. We decided to start with a simple and dynamic algorithm: evenly-balancing workers.
 
 ##### Evenly balanced workers
+
+The purpose of this algorithm is to attribute a fair amount of workers between each queue by dynamically assigning a queue to a worker, every second. This way if a queue is empty the workers that were polling from it can be evenly redistributed between the non-empty queues.
+
+Once we had a working implementation of the algorithm we decided to benchmark with the same scenario as we did for Sidekiq:
+
+![efficiency_algorithms_benchmark](/images/efficiency_algorithms_benchmark_2.png){:width="400"}
+
+The algorithm performed poorly compared to Sidekiq's random polling. It took 357 seconds to perform the same amount and type of jobs with our algorithm. This is because in the context of having CPU-blocking jobs, having multiple workers would not make a difference as compared to having one woker (when using MRI).
+
+How can Workerholic perform better?
+
 ##### Adaptive and Successive Algorithm (ASP)
+###### Iteration 1
+
+We decided to use a different, custom, algorithm: Adaptive and Successive Provisioning of workers based on queue load. This would allow workers to switch faster to IO queues in case there are any because CPU-bound are usually executed a faster rate than IO blocking jobs. Once again, we benchmarked using the same amount and type of jobs:
```

Comment:
We decided to use a different algorithm: we called it adaptive and successive provisioning of workers based on queue load. This would allow workers to switch faster to IO queues in case there are any because CPU-bound jobs are usually executed at a faster rate than IO blocking jobs. Once again we benchmarked using the same amount and type of jobs